File: server_integration_tests/java/java_server/src/main/java/dev/graphlink/javaserver/impl/NestingServiceImpl.java
package dev.graphlink.javaserver.impl;

import dev.graphlink.javaserver.generated.enums.Color;
import dev.graphlink.javaserver.generated.services.NestingService;
import dev.graphlink.javaserver.generated.types.Box;
import dev.graphlink.javaserver.generated.types.Circle;
import dev.graphlink.javaserver.generated.interfaces.Shape;
import dev.graphlink.javaserver.generated.types.Square;
import org.springframework.stereotype.Service;

import java.util.List;

/** List-depth (1/2/3) × nullability × kind (type / interface / enum) resolvers. */
@Service
public class NestingServiceImpl implements NestingService {

    private Box box(String id) {
        return Box.builder().id(id).label("label-" + id).build();
    }

    private Circle circle(String id) {
        return Circle.builder().id(id).kind("circle").radius(1).build();
    }

    private Square square(String id) {
        return Square.builder().id(id).kind("square").side(2).build();
    }

    @Override
    public List<Color> colors1() {
        return List.of(Color.RED, Color.GREEN, Color.BLUE);
    }

    @Override
    public List<? extends List<Color>> colors2() {
        return List.of(List.of(Color.RED), List.of(Color.GREEN, Color.BLUE));
    }

    @Override
    public List<? extends List<? extends List<Color>>> colors3() {
        return List.of(List.of(List.of(Color.RED)));
    }

    @Override
    public List<Box> boxes1() {
        return List.of(box("b1"), box("b2"));
    }

    @Override
    public List<? extends List<Box>> boxes2() {
        return List.of(List.of(box("b1")), List.of(box("b2")));
    }

    @Override
    public List<? extends List<? extends List<Box>>> boxes3() {
        return List.of(List.of(List.of(box("b1"))));
    }

    @Override
    public List<? extends Shape> shapes1() {
        return List.of(circle("c1"), square("s1"));
    }

    @Override
    public List<? extends List<? extends Shape>> shapes2() {
        return List.of(List.of(circle("c1")), List.of(square("s1")));
    }

    @Override
    public List<? extends List<? extends List<? extends Shape>>> shapes3() {
        return List.of(List.of(List.of(circle("c1"), square("s1"))));
    }
}

File: server_integration_tests/kotlin/kotlin_server_suspend/src/main/kotlin/dev/graphlink/kotlinserversuspend/impl/NestingServiceImpl.kt
package dev.graphlink.kotlinserversuspend.impl

import dev.graphlink.kotlinserversuspend.generated.enums.Color
import dev.graphlink.kotlinserversuspend.generated.interfaces.Media
import dev.graphlink.kotlinserversuspend.generated.interfaces.Shape
import dev.graphlink.kotlinserversuspend.generated.services.NestingService
import dev.graphlink.kotlinserversuspend.generated.types.Box
import dev.graphlink.kotlinserversuspend.generated.types.Circle
import dev.graphlink.kotlinserversuspend.generated.types.Photo
import dev.graphlink.kotlinserversuspend.generated.types.Square
import dev.graphlink.kotlinserversuspend.generated.types.Video
import org.springframework.stereotype.Service

/** List-depth (1/2/3) × nullability × kind (type / interface / union / enum) resolvers. */
@Service
class NestingServiceImpl : NestingService {

    private fun box(id: String) = Box(id = id, label = "label-$id")
    private fun circle(id: String) = Circle(id = id, kind = "circle", radius = 1)
    private fun square(id: String) = Square(id = id, kind = "square", side = 2)
    private fun photo(id: String) = Photo(id = id, url = "https://x/$id.jpg", width = 640)
    private fun video(id: String) = Video(id = id, url = "https://x/$id.mp4", durationSec = 30)

    override suspend fun colors1(): List<Color> = listOf(Color.RED, Color.GREEN, Color.BLUE)

    override suspend fun colors2(): List<List<Color?>?>? = listOf(listOf(Color.RED), listOf(Color.GREEN, Color.BLUE))

    override suspend fun colors3(): List<List<List<Color>>> = listOf(listOf(listOf(Color.RED)))

    override suspend fun boxes1(): List<Box?>? = listOf(box("b1"), box("b2"))

    override suspend fun boxes2(): List<List<Box>> = listOf(listOf(box("b1")), listOf(box("b2")))

    override suspend fun boxes3(): List<List<List<Box?>?>?>? = listOf(listOf(listOf(box("b1"))))

    override suspend fun shapes1(): List<Shape> = listOf(circle("c1"), square("s1"))

    override suspend fun shapes2(): List<List<Shape?>?>? = listOf(listOf(circle("c1")), listOf(square("s1")))

    override suspend fun shapes3(): List<List<List<Shape>>> = listOf(listOf(listOf(circle("c1"), square("s1"))))

    override suspend fun media1(): List<Media> = listOf(photo("p1"), video("v1"))

    override suspend fun media2(): List<List<Media?>?>? = listOf(listOf(photo("p1")), listOf(video("v1")))

    override suspend fun media3(): List<List<List<Media>>> = listOf(listOf(listOf(photo("p1"), video("v1"))))
}
